### PR TITLE
Ticket reference checks

### LINKED ISSUE #106: Ticket reference checks
Main obstacle: Tickets that do not live in same Github installation. Not the common case, but might happen.

### PR DESCRIPTION
**Pull request by prayerslayer**
Fixes #106 

* Patterns configurable via `commit.message.patterns`
* Check context is `zappr-commitmessage`, so it will appear different from `zappr` which is used for approvals
* All messages must match at least one pattern

TODO:
- [x] consider merge commits (commits w/ > 1 parent), they usually have messages like "Merge branch X into Y"

### CODE REVIEW

File: server/checks/CommitMessage.js [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+ */
+export function getAnyMatcherFn(regexes) {
+  return function (string) {
+    return regexes.reduce((result, regex) => result || regex.test(string), false)
+  }
+}
+
+function isMergeCommit({parents}) {
+  return Array.isArray(parents) && parents.length > 1
+}
```
Comment (mfellner) [inline server/checks/CommitMessage.js]: Please add JSDoc.

File: common/util.js [hunk 2]
```diff
@@ -23,3 +9,16 @@ export function promiseFirst(promises) {
     tryResolve(promises, resolve, reject)
   )
 }
+
+export function getIn(obj, path, returnDefault = null) {
```
Comment (mfellner) [inline common/util.js]: Please add JSDoc.

File: server/checks/index.js [hunk 3]
```diff
@@ -19,3 +22,4 @@ export function getCheckByType(type) {
 
 export { default as Approval } from './Approval'
 export { default as Autobranch } from './Autobranch'
+export { default as CommitMessage } from './CommitMessage'
```
Comment (mfellner) [inline server/checks/index.js]: Please add a new line

File: server/service/GithubService.js [hunk 4]
```diff
@@ -234,4 +236,17 @@ export default class GithubService {
     }
     return repos
   }
+
+  async fetchPullRequestCommits(owner, repo, number, accessToken) {
```
Comment (mfellner) [inline server/service/GithubService.js]: Please add JSDoc

File: test/common/util.test.js [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+    it('should return deep property', () => {
+      expect(getIn(obj, ['commit', 'message', 'patterns'])).to.deep.equal(obj.commit.message.patterns)
+    })
+
+    it('should return default value if path does not exist', () => {
+      expect(getIn(obj, ['commit', 'bar', 'patterns'], false)).to.be.false
+    })
+  })
+})
```
Comment (mfellner) [inline test/common/util.test.js]: Please add a newline.

File: test/server/commitmessage.test.js [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+        expect(github.setCommitStatus.callCount).to.equal(2)
+        expect(github.setCommitStatus.args.map(a => a[3].state)).to.deep.equal(['pending', 'failure'])
+        expect(github.setCommitStatus.args[1][3].description).to.equal('Commits 1commi, 3commi do not match configured patterns.')
+        done()
+      } catch(e) {
+        done(e)
+      }
+    })
+  })
+})
```
Comment (mfellner) [inline test/server/commitmessage.test.js]: Please add an emoji 🎱 test

**Comment by spaudanjo** [comment]
looks cool :) 
waiting for updates...

File: server/service/GithubService.js [hunk 7]
```diff
@@ -30,6 +32,8 @@ function fromBase64(encoded) {
 export default class GithubService {
 
   getOptions(method, path, body, accessToken) {
+    let url = nconf.get('GITHUB_URL') + path
+
```
Comment (mfellner) [inline server/service/GithubService.js]: This variable is not used! Please remove.

File: client/components/RepositoryCheck.jsx [hunk 8]
```diff
@@ -4,22 +4,16 @@ import { Alert, Panel, Well } from 'react-bootstrap'
...
             </p>,
   autobranch: <p>When enabling automatic branch creation, ZAPPR will automatically
-                 create a branch in your repository for every opened issue. By
-                 default the branch name will be <code>{'{issue.number}-{issue.title}'}</code>
-                 and maximum 60 characters long, but you can configure this in your
-                 <code>.zappr.yml</code>. Note that non-ASCII characters and
-                 punctuation will be removed.
-              </p>
+                 create a branch in your repository for every opened issue. For configuration options please consult
+                 <a href='https://zappr.readthedocs.org/en/latest/setup/#autobranch'>our docs</a>
```
Comment (mfellner) [inline client/components/RepositoryCheck.jsx]: There's a space missing between "consult" and "our docs". But we have the same text 3 times. Can we just make the check/feature name, e.g. "automatic branch creation" into a link? Would be cleaner, I think.

Comment (prayerslayer) [inline client/components/RepositoryCheck.jsx]: Maybe a link in parentheses to the docs after the check name? `... automatic branch creation (_docs_) ...`

File: server/checks/CommitMessage.js [hunk 9]
```diff
@@ -0,0 +1,102 @@
+import Check from './Check'
+import { logger } from '../../common/debug'
+import { getIn } from '../../common/util'
+import * as EVENTS from '../model/GithubEvents'
+
+const CHECK_TYPE = 'commitmessage'
+const info = logger(CHECK_TYPE, 'info')
+const error = logger(CHECK_TYPE, 'error')
+const context = 'zappr-commitmessage'
```
Comment (mfellner) [inline server/checks/CommitMessage.js]: Could we maybe rename this to `zappr/commitmessage` and the context of the Approval Check to `zappr/approval`?

**Comment by mfellner** [comment]
:+1:

**Pull request merged**